staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1365,29 +1364,111 @@ func hasServedCRDVersion(spec *apiextensionsv1.CustomResourceDefinitionSpec, ver
+		},
 	}
 
-	specs := []*spec.Swagger{}
 	for _, v := range crd.Spec.Versions {
 		// Defaults are not pruned here, but before being served.
 		// See flag description in builder.go for flag usage
 		s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true, SkipFilterSchemaForKubectlOpenAPIV2Validation: true, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
 		if err != nil {
+			// Failed to build open api
 			return nil, err
 		}
-		specs = append(specs, s)
+
+		// Build a dependency tree of which static definitions were referred by
+		// the spec
+		for k, def := range s.Definitions {
+			mergedOpenAPI.Definitions[k] = def
+		}
 	}
 
-	mergedOpenAPI, err := builder.MergeSpecs(staticOpenAPISpec, specs...)
+	//TEMPORARY HACK:
+	// Gnostic ParseDocument errors if there are unresolved refs. We don't want
+	// this behavior since some of the refs are found in the static type converter
+	//
+	// Workaround by inserting an empty schema for every type contained by
+	// static type converter.
+	//
+	// After parsing, we can remove the fake schemas.
+	// This will not be necessary once we remove gnostic ParseDocument by
+	// implementing direct conversion from spec.Schema to SMD Schema
+	toRemove := map[string]struct{}{}
+	gvkStaticTypes, ok := staticTypeConverter.(fieldmanager.LazyTypeConverter).TypeResolver().(*managedfields.GvkParser)
+
+	if ok {
+		for _, name := range gvkStaticTypes.Parser().TypeNames() {
+			if _, exists := mergedOpenAPI.Definitions[name]; !exists {
+				toRemove[name] = struct{}{}
+				mergedOpenAPI.Definitions[name] = spec.Schema{
+					SchemaProps: spec.SchemaProps{
+						Type: spec.StringOrArray{"object"},
+					},
+				}
+			}
+		}
+	} else {
+		return nil, errors.New("expected static type converter to use gvkParser")
+	}
+
+	// Implementation of ToProtoModels duplicated so we can remove fake schemas
+	//TEMPORARY HACK until direct conversion from spec.Schema -> TypeConverter
+	// is implemented
+	specBytes, err := gjson.MarshalIndent(mergedOpenAPI, " ", " ")
 	if err != nil {
 		return nil, err
 	}
-	models, err := utilopenapi.ToProtoModels(mergedOpenAPI)
+
+	doc, err := openapi_v2.ParseDocument(specBytes)
 	if err != nil {
 		return nil, err
 	}
-	return models, nil
+
+	filteredDefs := []*openapi_v2.NamedSchema{}
+
+	for _, def := range doc.Definitions.AdditionalProperties {
+		if _, shouldRemove := toRemove[def.Name]; !shouldRemove {
+			filteredDefs = append(filteredDefs, def)
+		}
+	}
+	doc.Definitions.AdditionalProperties = filteredDefs
+
+	models, err := proto.NewOpenAPIData(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	crdParser, err := managedfields.NewGVKParser(models, crd.Spec.PreserveUnknownFields)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add knowledge of static types not contained CRD schema to the parser
+	//!TODO: would it be better to add delegating functionalty to smdschema?
+	crdParser.AddModelsFromParser(gvkStaticTypes)
```

Comment:
Initially I thought it was just for the TypeMeta/Object meta references injected by the builder, so I had an earlier implementation which would hardcode just those two types. Eventually I switched to this design since there were cases where other types from autoscaling being needed, which also required more apimachinery dependencies like resource.Quantity.